src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3LinuxDriver.cs

```diff
@@ -367,6 +367,83 @@ private void SetInputPullModePi4(int pinNumber, PinMode mode)
+        {
+            Initialize();
+            /*
+             * There are 6 registers (4-byte ints) that control the mode for all pins. Each
+             * register controls the mode for 10 pins. Each pin uses 3 bits in the register
+             * containing the mode.
+             */
+
+            // Define the shift to get the right 3 bits in the register
+            int shift = (pinNumber % 10) * 3;
+            // Gets a pointer to the register that holds the mode for the pin
+            uint* registerPointer = &_registerViewPointer->GPFSEL[pinNumber / 10];
+            uint register = *registerPointer;
+            // Clear the 3 bits to 0 for the pin Number.
+            register &= ~(0b111U << shift);
+            // Set the 3 bits to the desired mode for that pin.
+            uint modeBits = 0; // Default: Gpio input
+
+            if (altMode == AlternatePinMode.Gpio)
+            {
+                // When setting back to Gpio, set to input, as this is the default
+                modeBits = 0;
+            }
+            else
+            {
+                modeBits = altMode.ModeValue;
+            }
+
+            register |= (modeBits) << shift;
+            *registerPointer = register;
+        }
+
+        /// <inheritdoc />
+        protected internal override AlternatePinMode GetAlternatePinMode(int pinNumber)
+        {
+            Initialize();
+            /*
+             * There are 6 registers(4-byte ints) that control the mode for all pins. Each
+             * register controls the mode for 10 pins. Each pin uses 3 bits in the register
+             * containing the mode.
+             */
+
+            // Define the shift to get the right 3 bits in the register
+            int shift = (pinNumber % 10) * 3;
+            // Gets a pointer to the register that holds the mode for the pin
+            uint* registerPointer = &_registerViewPointer->GPFSEL[pinNumber / 10];
+            uint register = *registerPointer;
+            // get the three bits of the register
+            register = (register >> shift) & 0b111;
+
+            switch (register)
+            {
+                case 0b000:
+                    // Input
+                    return AlternatePinMode.Gpio;
+                case 0b001:
+                    return AlternatePinMode.Gpio;
+                case 0b100:
+                    return AlternatePinMode.Alt0;
+                case 0b101:
+                    return AlternatePinMode.Alt1;
+                case 0b110:
+                    return AlternatePinMode.Alt2;
+                case 0b111:
+                    return AlternatePinMode.Alt3;
+                case 0b011:
+                    return AlternatePinMode.Alt4;
+                case 0b010:
+                    return AlternatePinMode.Alt5;
+            }
+
+            // This cannot happen.
+            throw new InvalidOperationException("Invalid register value");

```

Comment:
I wonder if we should just assume that if we can't read it then it's GPIO

Reply:
This here cannot realistically happen. The value is masked above to 0b111 and all cases are handled.

Reply:
technically you can add extra fields in the AlternatePinMode but assuming this actually shouldn't ever happen we shouldn't try to fix this anyway (I missed the comment previously)

Reply:
Hmm... True. But that applies to the inverse only, not this method. I'll add a check there.